katdal/test/conftest.py

```diff
@@ -0,0 +1,38 @@
+import pytest
+
+TEST_DURATION_TOLERANCE = 0.1
+
+
+def pytest_addoption(parser):
+    parser.addoption(
+        "--check-durations",
+        action="store_true",
+        help="Verify how long some tests run (the ones with an `expected_duration` mark)",
+    )
+
+
+@pytest.hookimpl(hookwrapper=True)
+def pytest_runtest_makereport(item, call):
+    """Optionally override pytest report creation to verify test duration."""
+    report = (yield).get_result()
+    # Only continue if the user requests this and the test has an expected_duration mark
+    check_durations = item.config.getoption("--check-durations", default=False)
+    mark = item.get_closest_marker("expected_duration")
+    if not check_durations or mark is None:
+        return report
+    # The test will take at least as long as the expected duration and probably a bit longer
+    minimum = mark.args[0]
+    maximum = minimum + TEST_DURATION_TOLERANCE
+    # Only verify duration if the test itself passes (and we are in the 'call' phase of test)
+    if (
+        report.when == 'call'
+        and report.passed
+        and not minimum <= report.duration <= maximum
+    ):
+        time_range = [minimum, maximum]
+        # Mark test as failed and report the timing discrepancy
+        report.outcome = 'failed'
+        report.longrepr = f"""
+        Test took {report.duration:g} seconds, which is outside the range {time_range}
+        """
```

Comment:
Does pytest do sensible things with the whitespace in the triple-quoted string?

Reply:
It looks like this:
```
_______________________________________ TestS3ChunkStoreToken.test_handle_read_paused_within_npy_array ________________________________________

        Test took 0.122755 seconds, which is outside the range [0.1, 0.11]
        
------------------------------------------------------------ Captured stdout setup ------------------------------------------------------------
```
So I guess it leaves it as is. I don't mind though.

Or do you think it's more standard to move the string to the start of the line?

Reply:
That's certainly not awful. What if you just made it a one-liner string with single quotes? Does it look okay without the embedded newlines?

Reply:
I remember it looked like this originally:
```
_______________________________________ TestS3ChunkStoreToken.test_handle_read_paused_within_npy_header _______________________________________
Test took 0.121048 seconds, which is outside the range [0.1, 0.11]
------------------------------------------------------------ Captured stdout setup ------------------------------------------------------------
```
Compared to the usual pytest blowout of a trace for a single error, this seemed too compact and easy to miss to me :-)

Plus I was trying to keep the line length down...

Reply:
OK, tweaked it (and lost two more lines of code :-))

Reply:
I'm happy with that.